### ISSUE TITLE
list<foo> in stubs

### ISSUE THREAD

**Issue by orklah** [vimeo/psalm#2534]
Hi, 

I noticed there isn't any list<foo> notation in the callmap file. However, I think most native functions actually return list<foo> instead of array<int,foo>

Is this by design? If not, is there a point in trying to submit a PR on this?

**Comment by muglug** [comment]
I'm ok with it if @TysonAndre is. Some of the most commonly-used functions like `explode` have their own return type providers that explicitly return lists: https://github.com/vimeo/psalm/blob/master/src/Psalm/Internal/Provider/ReturnTypeProvider/ExplodeReturnTypeProvider.php

**Comment by TysonAndre** [comment]
Phan's already using the `list<...>` syntax (and the `associative-array` syntax suggested as an enhancement to Psalm) in its version of the call map.

```
'array_chunk' => ['list<array>', 'input'=>'array', 'size'=>'int', 'preserve_keys='=>'bool'],
'array_keys' => ['list<string>|list<int>', 'input'=>'array', 'search_value='=>'mixed', 'strict='=>'bool'],
'array_values' => ['list<mixed>', 'input'=>'array'],
'call_user_func_array' => ['mixed|false', 'function'=>'callable', 'parameters'=>'list<mixed>'],
'call_user_method_array' => ['mixed', 'method_name'=>'string', 'obj'=>'object', 'params'=>'list<mixed>'],
'debug_backtrace' => ['list<array>', 'options='=>'int|bool', 'limit='=>'int'],
'dns_get_record' => ['list<array>|false', 'hostname'=>'string', 'type='=>'int', '&w_authns='=>'array', '&w_addtl='=>'array', 'raw='=>'bool'],
'explode' => ['non-empty-list<string>', 'separator'=>'string', 'str'=>'string', 'limit='=>'int'],
'fgetcsv' => ['list<?string>|false', 'fp'=>'resource', 'length='=>'int', 'delimiter='=>'string', 'enclosure='=>'string', 'escape='=>'string'],
'file' => ['list<string>|false', 'filename'=>'string', 'flags='=>'int', 'context='=>'resource'],
'forward_static_call_array' => ['mixed|false', 'function'=>'callable', 'parameters'=>'list<mixed>'],
'fscanf' => ['list<mixed>', 'stream'=>'resource', 'format'=>'string'],
'func_get_args' => ['list<mixed>'],
'get_class_methods' => ['list<string>', 'class'=>'mixed'],
'get_declared_classes' => ['list<class-string>'],
'get_declared_interfaces' => ['list<class-string>'],
'get_declared_traits' => ['list<class-string>'],
'get_extension_funcs' => ['list<callable-string>', 'extension_name'=>'string'],
'get_included_files' => ['list<string>'],
'get_loaded_extensions' => ['list<string>', 'zend_extensions='=>'bool'],
'gethostbynamel' => ['list<string>|false', 'hostname'=>'string'],
'getopt' => ['array<string,string>|array<string,false>|array<string,list<mixed>>', 'options'=>'string', 'longopts='=>'array', '&w_optind='=>'int'],
'glob' => ['list<string>|false', 'pattern'=>'string', 'flags='=>'int'],
'gzfile' => ['list<string>', 'filename'=>'string', 'use_include_path='=>'int'],
'hash_algos' => ['list<string>'],
'hash_hmac_algos' => ['list<string>'],
'headers_list' => ['list<string>'],
'mb_detect_order' => ['bool|list<string>', 'encoding_list='=>'mixed'],
'mb_encoding_aliases' => ['list<string>|false', 'encoding'=>'string'],
'mb_split' => ['list<string>', 'pattern'=>'string', 'string'=>'string', 'limit='=>'int'],
'ob_list_handlers' => ['list<string>'],
'openssl_get_curve_names' => ['list<string>'],
'preg_split' => ['list<string>', 'pattern'=>'string', 'subject'=>'string', 'limit='=>'?int', 'flags='=>'int'],
'ReflectionClass::getInterfaceNames' => ['list<class-string>'],
'ReflectionClass::getMethods' => ['list<ReflectionMethod>', 'filter='=>'int'],
'ReflectionClass::getProperties' => ['list<ReflectionProperty>', 'filter='=>'int'],
'ReflectionClass::getReflectionConstants' => ['list<ReflectionClassConstant>'],
'ReflectionClass::getTraitNames' => ['list<string>|null'],
'ReflectionClass::newInstanceArgs' => ['object', 'args='=>'list<mixed>'],
'ReflectionExtension::getClassNames' => ['list<string>'],
'ReflectionFunction::getParameters' => ['list<ReflectionParameter>'],
'ReflectionFunctionAbstract::getParameters' => ['list<ReflectionParameter>'],
'ReflectionMethod::getParameters' => ['list<\ReflectionParameter>'],
'ReflectionObject::getReflectionConstants' => ['list<\ReflectionClassConstant>'],
'ReflectionObject::getTraitNames' => ['list<string>'],
'scandir' => ['list<string>|false', 'dir'=>'string', 'sorting_order='=>'int', 'context='=>'resource'],
'sscanf' => ['list<mixed>|int', 'str'=>'string', 'format'=>'string', '&...w_vars='=>'string|int|float'],
'str_split' => ['list<string>', 'str'=>'string', 'split_length='=>'int'],
'stream_get_transports' => ['list<string>'],
'stream_get_wrappers' => ['list<string>'],
'timezone_identifiers_list' => ['list<string>|false', 'what='=>'int', 'country='=>'?string'],
'token_get_all' => ['list<string|array{0:int,1:string,2:int}>', 'source'=>'string', 'flags='=>'int'],
```

**Comment by muglug** [comment]
Oh ok – I've just added non-support for `associative-array` (made it an alias of `array`) in 79be178,  so if you want to add those changes to `phan/PHPSignatures`, go ahead!

**Comment by orklah** [comment]
@muglug What does that imply? Should I start working on a PR?

**Comment by muglug** [comment]
@orklah I'd wait 24 hours – if @TysonAndre can update https://github.com/phan/PHPSignatures with Phan's latest, I can then sync those changes into https://github.com/psalm/PHPSignatures before merging them into the CallMap, saving you the time.

**Comment by orklah** [comment]
Ok, will do!

**Comment by muglug** [comment]
Also @TysonAndre FYI `explode` can return an empty list if the given limit is negative

**Comment by TysonAndre** [comment]
> If the limit parameter is negative, all components except the last -limit are returned.

Huh, did not know that.

**Comment by orklah** [comment]
closed by #2675
